### PR TITLE
docs: import export for project/env

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

Documentation for the new environment import/export.

Important points:
* I tried to make it clear how it differs from the existing import/export

### CODE REVIEW

File: website/docs/reference/deploy/environment-import-export.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+The following items will be exported and imported:
+* feature toggles
+* activation strategies
+* variants
+* enabled/disabled statuses
+* context fields
+* feature tags
+* feature tag types
+* segments
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:25]: I think we could clarify already at this point that exporting features also exports global reference items etc.

we should also add doc links for as many items as we can on this list (I've not done that in the suggestion)

Here's is a rough suggestion for what I think would be good here. What do you think?

```suggestion
When you export features, the export will contain both feature-specific configuration and global configuration.

On the project-level these  items are exported:

* the feature itself
* feature tags

On the environment-level, these items are exported for the chosen environment:

* activation strategies including constraints and references to segments
* variants
* enabled/disabled

Additionally, these global configuration items are exported:
* custom context fields 
* feature tag types

Importantly, while references to segments are exported, the segments themselves are **not** exported. Consult the [import rejection](#import-rejection) section for more information.
```

File: website/docs/reference/deploy/environment-import-export.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+## Export
+
+Export is search-driven based on the results you see in the project toggle search or a global toggle search.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:29]: As a reader: what does search-driven mean? Is it something like this:

```suggestion
You can export features either from a project search or the global feature search. Use the search functionality to narrow the results to the list of features you want to export and use the export button to select environment. All features included in your search results will be included in the export.
```

But still, this feels very UI-centric and may be more at home in a how-to guide. How does the API work?

File: website/docs/reference/deploy/environment-import-export.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+![Search-driven export screen with WYSIWYG mode](/img/export.png)
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:31]: What is this image supposed to tell the reader? What information are they supposed to get from it? Just thinking the alt text could be better. And if it's a lot of words, consider using the `Figure` element and giving it a caption (check out `website/docs/topics/the-anatomy-of-unleash.mdx` for how to import and use the element).

File: website/docs/reference/deploy/environment-import-export.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+## Import
+
+Import is a 3 stage process designed to be efficient and error-resistant.
+
+### Import stages
+
+* **upload** - you can upload exported file or copy-paste export data into the code editor
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:39]: What's the file format? what about copy-pasted data? Must it be JSON or can it be ... XML? TOML? Where do they get this data? Is that copied from the exported file?

File: website/docs/reference/deploy/environment-import-export.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+* **validation** - you will get feedback on any errors or warnings before you do the actual import. This makes sure your feature flags configurations
+are compatible with your target environment.
+* **import** - the actual import that creates a new configuration in the target environment or creates a [change request](../change-requests.md) when the environment has change requests enabled
+
+![Three-staged import process](/img/import.png)
+
+### Import rejection
+
+Import can be rejected when:
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:48]: "Can" or "will"? 
```suggestion
An import will be rejected if:
```

File: website/docs/reference/deploy/environment-import-export.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+* the exported and imported context fields have conflicting context fields values. If the context fields don't exist in the target environment they will be created.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:49]: > the exported and imported context fields have conflicting context fields values.

What does this mean?

File: website/docs/reference/deploy/environment-import-export.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+* the target Unleash instance doesn't have a matching segment. The actual constraints in the segment can be different. The matching is done by the segment name.
+* the exported strategies include custom strategies that don't exist in the target Unleash instance. The matching is done by the strategy name.
+* the exported feature already exists in another project in the target environment that is not a project that you import the toggle to.
+* the target project and environment have a pending change request (applies to enterprise customers with change requests enabled for a given project and environment).
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:53]: I think each list item here can be its own subsection so that we can explain it in full without overcrowding the list. There is a lot of nuance to some (most?) of the items and I don't think the list format serves that purpose very well.

File: website/docs/reference/deploy/environment-import-export.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+![Three-staged import process](/img/import.png)
+
+### Import rejection
+
+Import can be rejected when:
+* the exported and imported context fields have conflicting context fields values. If the context fields don't exist in the target environment they will be created.
+* the target Unleash instance doesn't have a matching segment. The actual constraints in the segment can be different. The matching is done by the segment name.
+* the exported strategies include custom strategies that don't exist in the target Unleash instance. The matching is done by the strategy name.
+* the exported feature already exists in another project in the target environment that is not a project that you import the toggle to.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:52]:
```suggestion
* you're trying to import a feature that already exists in a different project on the target instance.
```

File: website/docs/reference/deploy/environment-import-export.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+* the target project and environment have a pending change request (applies to enterprise customers with change requests enabled for a given project and environment).
+
+### Import warnings
+
+Import validation will warn you when:
+* the features you are trying to import are already archived in the target instance. Those features won't be imported again.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:58]: This is regardless of the what project they are archived from? If you're trying to import feature F into project P, but a feature called F has already been archived in project Q: what happens? Do you get an error or a warning?

File: website/docs/reference/deploy/environment-import-export.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+* the strategies you are trying to import are custom and may have different configuration parameters. Those strategies will be imported anyway.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:59]: > the strategies you are trying to import are custom and may have different configuration parameters.

What does it mean that they **may** have different parameters? Do you get this warning whenever you import custom strategies or only when you import custom strategies that have different parameters?

File: website/docs/reference/deploy/environment-import-export.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+### Required permissions
+
+Import requires those permissions:
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:63]: Judging from the list (and from what we discussed previously), it doesn't always need all of these, right? Sometimes, it's just a few of them?

File: website/docs/reference/deploy/environment-import-export.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+* Update feature toggles (always)
+* Create feature toggles when import will be creating new features
+* Update tag types when import will be introducing new tag types
+* Create context fields when import will be creating new context fields
+* Create activation strategies when import will be creating new activation strategies and change requests are disabled
+* Delete activation strategies when import will be deleting existing activation strategies and change requests are disabled
+* Update variants when import will be updating variants and change requests are disabled
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:70]: What happens to the last three when change requests are enabled? Does enabling change requests change any of the other ones?

Also is create/delete activation strategies related to creating/deleting custom strategies or to adding/removing strategies from a feature?

```suggestion
To import features, you will **always** require the **update feature toggles** permission.

Additionally, depending on how the actions your import job would trigger, you may also require any of the following permissions:

* **Create feature toggles**: When the import would create new features
* **Update tag types**: when the import would create new tag types
* **Create context fields**: when the import would create new context fields
* **Create activation strategies**: when the import would add (or is this "create custom activation strategies"? custom or standard?) activation strategies to a feature and change requests are disabled
* **Delete activation strategies**: when import would remove existing activation strategies from a feature and change requests are disabled
* **Update variants**: when the import would update variants and change requests are disabled

When change requests are enabled ....
```

File: website/docs/reference/deploy/environment-import-export.md [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+### Import and change requests
+
+When the target project and environment have change requests enabled the import will not be applied instantly. Instead, it will 
+create a change request, send it to review and await for the approval.
+
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:76]: Maybe we could add a paragraph here explaining how it affects permissions?

```suggestion
If change requests are enabled, any permissions for .... are not required.
```

File: website/docs/reference/deploy/environment-import-export.md [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+## Environment Import & Export vs Old Import & Export
+
+The new Environment Import & Export differs from the old [Import & Export](import-export.md).
+The most important difference is a strong focus on a single project and environment in the target environment.
+Also, the new Import & Export has a UI and extra safety mechanisms preventing corrupted data imports.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.md:81]: This is nice and a good clarification. So when would you pick the new one over the old one? And maybe more interestingly, when would you pick the old one over the new one?

File: website/docs/reference/deploy/environment-import-export.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+It also makes it easier to move feature flags within a single instance between environments.
+When exporting you always select the toggles and environment to export from. When importing you always select a project and an environment to import to.
+
+## Import & Export items
+
+The following items will be exported and imported:
+* feature toggles
+* activation strategies
+* variants
+* enabled/disabled statuses
```
[the remark this answers is not in the document]
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.md:23]: yes

File: website/docs/reference/deploy/environment-import-export.md (hunk 2 above)
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.md:31]: This feature is UI centric as UI import is our main use case. I took the wording from our import/export article.

Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.md:31]: At the API level you have a list of feature IDs to export and the env name

File: website/docs/reference/deploy/environment-import-export.md (hunk 14 above)
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.md:83]: Hmm. This is on line 80 and 81. When you want to import into single project and environment. And if you want to have a UI and the validation stage.

The motivation for the old one should be in the docs of the old one I assume. From what I know it was trying to be too ambitious and export too much, but you'd need to talk to @ivarconr

File: website/docs/reference/deploy/environment-import-export.md (hunk 1 above)
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.md:27]: I like the extra clarity that it brings.

File: website/docs/reference/deploy/environment-import-export.md (hunk 6 above)
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.md]: Added clarification. Please let me know if it makes more sense now

File: website/docs/reference/deploy/environment-import-export.md (hunk 8 above)
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.md]: I like it

File: website/docs/reference/deploy/environment-import-export.md (hunk 9 above)
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.md]: Added a tiny clarification "in any project". Isn't unique toggle per unleash instance documented somewhere else? Import/export is not changing anything in this regard.

File: website/docs/reference/deploy/environment-import-export.md (hunk 10 above)
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.md]: Always when importing custom strategies. To remind you to check the strategy configuration if it happens to be part of our import.

Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.md]: Added a link to configuration parameters

File: website/docs/reference/deploy/environment-import-export.md (hunk 12 above)
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.md]: I like the suggestion

File: website/docs/reference/deploy/environment-import-export.md (hunk 13 above)
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.md]: Agree, Let's be explicit and add this here

File: website/docs/reference/unleash-context.md [hunk 16]
```diff
@@ -40,7 +40,7 @@ As an example: You've created a custom field called `groupId`. You know group ID
 
 
 
-## Custom context fields
+## Custom context fields {#context-fields}
```
Comment (thomasheartman) [inline website/docs/reference/unleash-context.md:43]: Out of curiosity: why did you change the id here? Have you checked whether that breaks any other links to this section from within the docs?

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+
+* **upload** - you can upload previously exported JSON file or copy-paste export data from the exported JSON file into the code editor
+* **validation** - you will get feedback on any errors or warnings before you do the actual import. This makes sure your feature flags configurations. You will not be able to finish the import if you have errors. Warnings don't stop you from importing.
+* **import** - the actual import that creates a new configuration in the target environment or creates a [change request](../change-requests.md) when the environment has change requests enabled
+
+![Three-staged import process](/img/import.png)
+
+### Import rejection
+
+An import will be rejected when there's a possibility of data corruption.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:58]: What does this mean? How do we identify data corruption? Or are we talking about any potential issues that may stop the import (as listed in the subsections)?

Like this?

```suggestion
Unleash will reject an import if it discovers conflicts between the data to be imported and what already exists on the Unleash instance. The import tool will tell you about why an import is rejected in these cases. 

The following sections describe requirements that must be met for Unleash not to stop the import job.
```

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+
+#### Context fields
+
+The exported and imported context fields have the same name but the target environment doesn't support some of the source environment context field values. If the context fields don't exist in the target environment they will be created.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:62]: I still don't understand this. What are the imported context fields and what are the exported context fields? Is this about conflicts between context fields that you are trying to import and context fields that already exist in the Unleash instance that you're trying to import them into?

I think there's three cases here, right? (Note: the "new" context field is the imported one, the "old" context field is the one that already exists in the instance)
1. You import a context field that doesn't exist in the target instance. All good.
2. You import a context field that already exists in the target instance. The new and the old context fields either have the exact same legal values or both accept everything. This is fine.
3. You import a context field that already exists in the target instance. There is a mismatch between the old and the new one. One of:
  i. The new context field has no restrictions on the value, but the existing one has a defined set of legal values
  ii. The old context field has no restrictions on the value, but the new field has a defined set of legal values
  iii. Both fields have legal values, but they are **not exactly the same** (even if the new custom context field's values are a strict subset of the old field's values)

Is that correct?

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+
+#### Segments
+
+The target Unleash instance doesn't have a matching segment. The actual constraints in the segment can be different. The matching is done by the segment name.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:66]:
```suggestion
If your import has segments, then a segment with the same name must already exist in the Unleash instance you are trying to import into. Only the name must be the same: the constraints in the segment can be different.
```

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+
+#### Custom strategies
+
+The exported strategies include custom strategies that don't exist in the target Unleash instance. The matching is done by the strategy name.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:70]:
```suggestion
If your import contains custom strategies, then custom strategies with the same names must already exist in the target Unleash instance. The custom strategy definitions (including strategy parameters) that exist in the target instance do not otherwise need to match the custom strategy definitions in the instance the import came from.
```

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+
+#### Existing features
+
+You're trying to import a feature that already exists in a different project on the target instance.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:74]:
```suggestion
If any of the features you import already exist in the target Unleash instance, then they must exist within the project you're importing into. If any features you are attempting to import exist in a **different** project on the target instance, the import will be rejected.
```

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+
+#### Pending change request
+
+The target project and environment have a pending change request (applies to enterprise customers with change requests enabled for a given project and environment).
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:78]:
```suggestion
#### Pending change requests

The project and environment you are importing into must **not have any pending [change requests](./change-requests.md)**. 
```

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+
+### Import warnings
+
+Import validation will warn you when:
+* the features you are trying to import are already archived in the target instance in any project. Those features won't be imported again.
+* the strategies you are trying to import are custom and may have different [configuration parameters](../custom-activation-strategies.md#parameters). Those strategies will be imported anyway.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:84]: So custom strategies always cause either a warning or an error, right? And why don't we validate custom strategies? Not worth the effort?

```suggestion
The import validation system will warn you about potential problems it finds in the import data. These warnings do not prevent you from importing the data, but you should read them carefully to ensure that Unleash does as you intend.

The following sections list things that the import tool will warn you about.

#### Archived features

The import tool will not import any features that have already been archived on the target Unleash instance. Because features are identified by their name, that means that if a feature called `feature-a` has been created and archived on the target Unleash instance, then a feature with the same name (`feature-a`) will not be imported.

If you permanently delete the archived `feature-a` from the target instance, then the new `feature-a` (in the import data) **will** be imported. 

#### Custom strategies

Unleash will verify that any custom strategies you are trying to import have already been defined on the target instance. However, it only does this verification by name. It does **not** validate that the definitions are otherwise the same or that they have the same [configuration parameters](../custom-activation-strategies.md#parameters). 

Before proceeding with the import, you should be sure that the custom strategy definitions are the same in the target instance and the imported custom strategies.
```

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+
+### Required permissions
+
+To import features, you will **always** require the **update feature toggles** permission.
+Additionally, depending on how the actions your import job would trigger, you may also require any of the following permissions:
+* **Create feature toggles**: When the import would create new features
+* **Update tag types**: when the import would create new tag types
+* **Create context fields**: when the import would create new context fields
+* **Create activation strategies**: when the import would add (or is this "create custom activation strategies"? custom or standard?) activation strategies to a feature and change requests are disabled
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:93]: I think you missed the question I added for you on this line 😅  Could you clarify whether this is for creating custom strategies or adding a strategy to a feature?

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+### Import warnings
+
+Import validation will warn you when:
+* the features you are trying to import are already archived in the target instance in any project. Those features won't be imported again.
+* the strategies you are trying to import are custom and may have different [configuration parameters](../custom-activation-strategies.md#parameters). Those strategies will be imported anyway.
+
+### Required permissions
+
+To import features, you will **always** require the **update feature toggles** permission.
+Additionally, depending on how the actions your import job would trigger, you may also require any of the following permissions:
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:89]: Whoops, an extra word slipped in there

```suggestion
Additionally, depending on the actions your import job would trigger, you may also require any of the following permissions:
```

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+* **Create feature toggles**: When the import would create new features
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:90]: All other lines start with a lowercase letter, so let's be consistent.

```suggestion
* **Create feature toggles**: when the import would create new features
```

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 27]
```diff
@@ -0,0 +1,110 @@
...
+* **Create context fields**: when the import would create new context fields
+* **Create activation strategies**: when the import would add (or is this "create custom activation strategies"? custom or standard?) activation strategies to a feature and change requests are disabled
+* **Delete activation strategies**: when import would remove existing activation strategies from a feature and change requests are disabled
+* **Update variants**: when the import would update variants and change requests are disabled
+
+
+### Import and change requests
+
+When the target project and environment have change requests enabled the import will not be applied instantly. Instead, it will
+create a change request, send it to review and await for the approval.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:101]: Something like this maybe? 

```suggestion 
If change requests are enabled for the target project and environment, the import will not be fully applied. Any new features will be created, but all feature configuration will be added to a new change request.
```

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 28]
```diff
@@ -0,0 +1,110 @@
...
+
+### Import stages
+
+* **upload** - you can upload previously exported JSON file or copy-paste export data from the exported JSON file into the code editor
+* **validation** - you will get feedback on any errors or warnings before you do the actual import. This makes sure your feature flags configurations. You will not be able to finish the import if you have errors. Warnings don't stop you from importing.
+* **import** - the actual import that creates a new configuration in the target environment or creates a [change request](../change-requests.md) when the environment has change requests enabled
+
+![Three-staged import process](/img/import.png)
+
+### Import rejection
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:56]: Maybe we can rephrase this header and section to be "import requirements" instead of import rejection?

```suggestion
### Import requirements
```

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+When the target project and environment have change requests enabled the import will not be applied instantly. Instead, it will
+create a change request, send it to review and await for the approval.
+
+If change requests are enabled, any permissions for  **Create activation strategies**, **Delete activation strategies** and **Update variants** are not required.
+
+## Environment Import & Export vs Old Import & Export
+
+The new Environment Import & Export differs from the old [Import & Export](import-export.md).
+The most important difference is a strong focus on a single project and environment in the target environment.
+Also, the new Import & Export has a UI and extra safety mechanisms preventing corrupted data imports.
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:109]: I don't have the time right now, but I'll get back to this later. I think it could do with some fleshing out.

**Review by thomasheartman — COMMENTED** [review]
Nice one! I think there's more improvements and clarifications we can make, though 😄

File: website/docs/reference/unleash-context.md (hunk 16 above)
Comment (kwasniew) [reply website/docs/reference/unleash-context.md]: I though it's not created automatically without this. Reverting this change as I see that's not the case and I can use #custom-context-field

File: website/docs/reference/deploy/environment-import-export.mdx (hunk 23 above)
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.mdx:84]: Because you may want the custom strategy work differently in those two instances. So we leave it up to you to decide if you're ok exporting a custom strategy. Also my impression is that we may want to start using constraints instead of custom strategies. So not worth investing into it too much right now.

Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.mdx:84]: Custom strategy gives a warning if you have it in a target instance. Custom strategy gives you an error if you don't have it in a target instance.

File: website/docs/reference/deploy/environment-import-export.mdx (hunk 18 above)
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.mdx:62]: The actual algorithms is much simpler: https://github.com/Unleash/unleash/blob/main/src/lib/features/export-import-toggles/import-context-validation.ts
Here's the test for it: https://github.com/Unleash/unleash/blob/main/src/lib/features/export-import-toggles/import-context-validation.test.ts

In essence, if you're importing legal values that don't exist in the target instance you get an error.

File: website/docs/reference/deploy/environment-import-export.mdx (hunk 17 above)
Comment (kwasniew) [reply website/docs/reference/deploy/environment-import-export.mdx:58]: YEs

File: website/docs/reference/deploy/environment-import-export.mdx (hunk 18 above)
Comment (thomasheartman) [reply website/docs/reference/deploy/environment-import-export.mdx:64]: Yeah, okay, so it's more like two:
1. The imported field doesn't exist already. All good.
2. If the imported context field has legal values, then the existing context field must also have at least those values defined. Any other values in the existing field is also okay.

File: website/docs/reference/deploy/environment-import-export.mdx (hunk 23 above)
Comment (thomasheartman) [reply website/docs/reference/deploy/environment-import-export.mdx:86]: Got it, thanks. In that case, I think the suggestion still stands, but we could maybe reword or remove the last section:

```suggestion
The import validation system will warn you about potential problems it finds in the import data. These warnings do not prevent you from importing the data, but you should read them carefully to ensure that Unleash does as you intend.

The following sections list things that the import tool will warn you about.

#### Archived features

The import tool will not import any features that have already been archived on the target Unleash instance. Because features are identified by their name, that means that if a feature called `feature-a` has been created and archived on the target Unleash instance, then a feature with the same name (`feature-a`) will not be imported.

If you permanently delete the archived `feature-a` from the target instance, then the new `feature-a` (in the import data) **will** be imported. 

#### Custom strategies

Unleash will verify that any custom strategies you are trying to import have already been defined on the target instance. However, it only does this verification by name. It does **not** validate that the definitions are otherwise the same or that they have the same [configuration parameters](../custom-activation-strategies.md#parameters). 
```

File: website/docs/reference/deploy/environment-import-export.mdx (hunk 24 above)
Comment (thomasheartman) [reply website/docs/reference/deploy/environment-import-export.mdx]: Okay. That's ... pretty confusing to me. But okay, I'll make a suggestion for it.

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 30]
```diff
@@ -0,0 +1,111 @@
...
+* the strategies you are trying to import are custom and may have different [configuration parameters](../custom-activation-strategies.md#parameters). Those strategies will be imported anyway.
+
+### Required permissions
+
+To import features, you will **always** require the **update feature toggles** permission.
+Additionally, depending on the actions your import job would trigger, you may also require any of the following permissions:
+* **Create feature toggles**: when the import would create new features
+* **Update tag types**: when the import would create new tag types
+* **Create context fields**: when the import would create new context fields
+* **Create activation strategies**: when the import would add an activation strategies to a feature and change requests are disabled
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:95]:
```suggestion
* **Create activation strategies**: when the import would add activation strategies to a feature and change requests are disabled
```

File: website/docs/reference/deploy/environment-import-export.mdx (hunk 29 above)
Comment (thomasheartman) [reply website/docs/reference/deploy/environment-import-export.mdx:110]: Maybe something like this?

```suggestion
## Environment import/export vs the [instance import/export API](import-export.md)

Environment import/export has some similarities to the [instance import/export API](import-export.md), but they serve different purposes. 

The instance import/export API was designed to export all feature toggles (optionally with strategies and projects) from one Unleash instance to another. When it was developed, Unleash had much fewer features than it does now. As such, the API lacks support for some of the more recent features in Unleash.

On the other hand, the environment import/export feature was designed to export a selection of features based on search criteria. It can only export data from a single environment and only import it to a single environment. It also only supports importing into a single project (although it can export features from multiple projects).

Further, the environment import/export comes with a much more stringent validation and will attempt to stop any corrupted data imports.
```

File: website/docs/reference/deploy/environment-import-export.mdx (hunk 18 above)
Comment (thomasheartman) [reply website/docs/reference/deploy/environment-import-export.mdx:64]:
```suggestion
When you import a custom context field **with legal values defined**:

If a custom context field with the same name already exists in the target instance, then the pre-existing context field must have **at least** those legal values defined. In other words, the imported context field legal values must be a subset of the already existing context field's legal values.

When you import custom context fields without legal values or custom context fields that don't already exist in the target instance, then there are no requirements. 

Custom context fields that don't already exist in the target instance will be created upon import.
```

File: website/docs/reference/deploy/environment-import-export.mdx [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+* [custom context fields](../unleash-context.md#custom-context-fields)
+* [feature tag types](../tags.md#tag-types)
+
+Importantly, while references to [segments](../segments.mdx)) are exported, the segments themselves are **not** exported. Consult the [import rejection](#import-rejection) section for more information.
+
+## Export
+
+You can export features either from a project search or the global feature search. Use the search functionality to narrow the results to the list of features you want to export and use the export button to select environment. All features included in your search results will be included in the export.
+
+<Figure caption="Search-driven export screen with WYSIWYG mode."  img="/img/export.png"/>
```
Comment (thomasheartman) [inline website/docs/reference/deploy/environment-import-export.mdx:42]:
```suggestion
<Figure caption="Feature toggle lists can be filtered using the search bar. Once filtered, you can use the "export current selection" button to export the configuration for those features."  img="/img/export.png"/>
```

**Review by thomasheartman — APPROVED** [review]
I've added my changes now as you requested. I'd like for you to look through the changes in 6f3fb38192d31146660c813f793d0c9f6a7fa71e before merging this, however, to make sure that the information presented there is correct. That goes especially for the section on importing custom context fields.

Once you're happy with that, I think we can go ahead and get this merged. Good job and thanks for sticking with me! 💯
